### ISSUE TITLE
Bugs in Metrics

### ISSUE THREAD

**Issue by Roffild** [dmlc/xgboost#6731]
gamma-nloglik:
```c++
bst_float psi = 1.0;
bst_float c = 1. / psi * std::log(y/psi) - std::log(y) - common::LogGamma(1. / psi);
```
***c == 0***

logloss:
```c++
  XGBOOST_DEVICE bst_float EvalRow(bst_float y, bst_float py) const {
    const bst_float eps = 1e-16f;
    const bst_float pneg = 1.0f - py;
    if (py < eps) {
      return -y * std::log(eps) - (1.0f - y)  * std::log(1.0f - eps);
    } else if (pneg < eps) {
      return -y * std::log(1.0f - eps) - (1.0f - y)  * std::log(eps);
    } else {
      return -y * std::log(py) - (1.0f - y) * std::log(pneg);
    }
  }
```
***std::log(1.0f - eps) == std::log(1.0) == 0***

gamma-deviance needs to be removed because the formula is not correct!(#6728)

I don't understand what math formula was used in poisson-nloglik.
-log( [Poisson_regression](https://en.wikipedia.org/wiki/Poisson_regression) )

With tweedie-nloglik it is also unclear. And the test is missing.

Tests for regression metrics with weights.(#6729)

If metrics are used in forest creation.....

**Comment by Roffild** [comment]
[I wanted to use metrics from XGBoost for Pytorch.](https://gist.github.com/Roffild/982ba8ffb1f7ee8a5f4f0183cbbf1cc0)

But now only [sklearn.metrics](https://scikit-learn.org/stable/modules/model_evaluation.html) for all models!

**Comment by trivialfis** [comment]
I need to take a closer look.

**Comment by trivialfis** [comment]
The bug in gamma deviance is fixed.  Better documentation for other metrics will be a different topic.  Thanks for raising the issue!

**Comment by trivialfis** [comment]
Just a quick note for who's following this thread.  I believe these metrics and objectives are derived from the generalized linear model.

**Comment by Roffild** [comment]
All metrics are calculated for each result separately.

Loss is calculated for each individual result, but metrics must be calculated for the entire result matrix. Therefore, the metrics in XGBoost are approximate.

**Comment by Roffild** [comment]
https://github.com/pytorch/pytorch/issues/22439

**Comment by trivialfis** [comment]
> I don't understand what math formula was used in poisson-nloglik.

See `Evaluating the Poisson distribution` section of https://en.wikipedia.org/wiki/Poisson_distribution

**Comment by trivialfis** [comment]
> gamma-nloglik:

Yeah, this one is a bit confusing, I tracked down the PR for `-log(\gamma)`: https://github.com/dmlc/xgboost/pull/1369 , which hard codes the dispersion to 1.  Not entirely sure why.

**Comment by trivialfis** [comment]
The weird logloss you see is just a way to work around numerical issues.
